### ISSUE TITLE
use unwrap_or_else + panic! instead of unwrap directly sometimes

### ISSUE THREAD

**Issue by siddontang** [pingcap/tikv#1376]
Now we use `unwrap` in many places, but when we really meet the panic, the stack is raised from rust lib result.rs, not the real panic line, like:

```
ERROR - thread 'main' panicked 'called `Result::unwrap()` on an `Err` value: "Invalid argument: db has more levels than options.num_levels"' at "../src/libcore/result.rs:788"
stack backtrace:
```

We can use unwrap_or_else + panic! to show the detailed error message and real panic file and line.
Of course, we don't need to replace all `unwrap`. Maybe we can start with `tikv-server.rs` first.

**Comment by BusyJay** [comment]
I don't get it, the stack trace should be printed out, which will reach the `unwrap` at some point. And the 'real' panic file and line do actually locate in result.rs, because `panic` is called at that point.

**Comment by mrmiywj** [comment]
Yep, but that not works on OSX. That’s the [issue of rust](https://github.com/rust-lang/rust/issues/24346 <https://github.com/rust-lang/rust/issues/24346>). It seems that they do not come out a plan to support it...

**Comment by BusyJay** [comment]
Use `unwrap_or_else` doesn't solve the problem too in this case, the line number is still missing. And normally our service will not be deployed on OS X.

**Comment by siddontang** [comment]
What I want most is the know the actual panic place when I first see the error, not `at "../src/libcore/result.rs:788"`. 

Using `unwrap`, I have to look the panic place through the stack, which is not convenient, and the stack looks horrible on OS X for development. 

Of course, using `unwrap_or_else` + `panic!` may cause the code longer compared with `unwrap`, but I think it is ok.

**Comment by BusyJay** [comment]
I propose to use `macro` in this case. For example define a `unwrap` macro to act as `Result::unwrap` exactly.

**Comment by hhkbp2** [comment]
IMO, it's the right thing to do to push Rust-lang team to solve rust-lang/rust#24346.
It takes heavy manual work to replace all these `unwrap` in this repo with `unwrap_or_else`/macro/or other code, which seems to try to solve this issue for MacOS in a clumsy way, even though MacOS is not the main platform where TiKV runs.

**Comment by siddontang** [comment]
As I said above, what I focus most is know the actual panic place directly, not `at "../src/libcore/result.rs:788"`.

**Comment by overvenus** [comment]
FYI, it seems that https://github.com/rust-lang/rfcs/pull/2091 will slove this issue.

**Comment by brson** [comment]
Tracking issue for upstream RFC: https://github.com/rust-lang/rust/issues/47809

Looks like @kennytm has been interested in it. There are fresh guidelines there on how to implement the feature since he last looked though.

**Comment by siddontang** [comment]
Great @kennytm
